pkg/detectors/dockerhub/dockerhub.go

```diff
@@ -0,0 +1,91 @@
+package dockerhub
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/trufflesecurity/trufflehog/v3/pkg/common"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/detectors"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/detectorspb"
+)
+
+type Scanner struct{}
+
+// Ensure the Scanner satisfies the interface at compile time.
+var _ detectors.Detector = (*Scanner)(nil)
+
+var (
+	client = common.SaneHttpClient()
+
+	// Can use email or username for login.
+	usernamePat = regexp.MustCompile(`(?im)(?:user|usr|-u)\S{0,40}?[:=\s]{1,3}[ '"=]{0,1}([a-zA-Z0-9]{4,40})\b[^@]?`)
+	emailPat    = regexp.MustCompile(common.EmailPattern)
+
+	// Can use password or personal access token (PAT) for login, but this scanner will only check for PATs.
+	accessTokenPat = regexp.MustCompile(`\b(?:dckr_pat_)([a-zA-Z0-9_-]){27}\b`)
+)
+
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"dckr_pat_"}
+}
+
+// FromData will find and optionally verify Dockerhub secrets in a given set of bytes.
+func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
+	dataStr := string(data)
+
+	usernameMatches := usernamePat.FindAllStringSubmatch(dataStr, -1)
+	emailMatches := emailPat.FindAllString(dataStr, -1)
+	accessTokenMatches := accessTokenPat.FindAllString(dataStr, -1)
+
+	userMatches := emailMatches
+	for _, usernameMatch := range usernameMatches {
+		userMatches = append(userMatches, usernameMatch[1])
```

Comment:
suggestion: We might want to check the len of `usernameMatch` here to prevent an IOOR panic.

For example if `usernamePat` matches a part of dataStr, but the match does not include any submatches. As a result, `usernameMatches` would be a slice of slice of strings where the inner slices have only 1 string.

We can include a quick check like:
```
if len(usernameMatch) > 1 {
        userMatches = append(userMatches, usernameMatch[1])
    }
```

Reply:
sounds good!